### DISCUSSION TITLE
Mapping configuration is not considered and constructor is used instead

### DISCUSSION THREAD

**Discussion by JoFrMueller** [LuckyPennySoftware/AutoMapper#3862] [NA]
### Source/destination types

The example is a console app that can be used 1:1.

### Mapping configuration

Example code contains mapping configuration.

### Version:

11.0.0

### Expected behavior

Mapping configuration should apply instead of "magic" constructor call.

### Actual behavior

The mapping configuration isn't considered at all.

### Steps to reproduce

```csharp
namespace ConsoleApp2
{
    class Program
    {
        static void Main(string[] args)
        {
            var config = new MapperConfiguration(cfg => {
                cfg.CreateMap<A, B>().ForMember(dest => dest.Credential, opt => opt.MapFrom(src => $"{src.Credential}|{src.Credential}"));
            });
            IMapper mapper = config.CreateMapper();
            var result = mapper.Map<B>(new A
            {
                Credential = "test"
            });

            if (result.Credential == "test")
            {
                throw new Exception("This was not expected... We expected test|test...");
            }
        }
    }
    
    class A
    {
        public string Credential { get; set; }
    }

    class B
    {
        public string Credential { get; private set; }

        public B()
        {
        }

        // If this constructor is removed, it behaves as expected.
        public B(string credential)
        {
            Credential = credential;
        }

        public void SetCredential(string credential)
        {
            Credential = credential;
        }
    }
}
```

Any further information or guidance on this?
In this example, the error is trivial to spot, but we had it in a more complex mapping configuration and we were desperately wondering what the rules exactly are, that decide if constructor parameter mapping or mapping configuration "wins"...

**Comment by lbargaoanu** [comment]
Things are not mapped twice, that's the rule :) Maybe just disable constructor mapping? There are many similar questions, here and on SO.

**Reply by JoFrMueller** [reply]
With your hint regarding "things are not mapped twice" and "disable constructor mapping" I think I've got it now. Thanks.

So constructor mapping wins over explicit configuration as long as the following is not called:

```csharp
var configuration = new MapperConfiguration(cfg => cfg.DisableConstructorMapping());
```

**Reply by ChristophHornung** [reply]
We ran into the same problem, where a developer added a new constructor in the same style in addition to the empty one which as a result broke the mapping.
If the rule is never to map twice can this rule be somehow enforced with validating or on building the configuration. (`AssertConfigurationIsValid` does not throw an exception e.g.)?

The way things stand the only reasonable approach to guard against a new constructor breaking the mappings (if you need constructor mappings on some entities) is to explicitly specify the constructor to be used for each mapping, which adds quite a lot of overhead.

**Reply by lbargaoanu** [reply]
You can write [your own](https://github.com/AutoMapper/AutoMapper/blob/bdc0120497d192a2741183415543f6119f50a982/src/UnitTests/CustomValidations.cs#L42) validation rules or use `ForAllMaps`. Unit tests also help :)
You can also configure what constructors AM uses with `ShouldUseConstructor`.

**Reply by ChristophHornung** [reply]
@lbargaoanu Thanks for the link to custom validation.

If anyone else is interested, I ended up adding this validation to our unit-tests, it checks for non-default mappings if the constructor would map the same property. At least for our quite large mapping all instances it found were misconfigurations, I don't know if it might miss some edge cases though.

```C#
private static void ValidateDuplicateMappings(ValidationContext validationContext)
	{
		if (validationContext.TypeMap?.ConstructorMap != null)
		{
			foreach (PropertyMap map in validationContext.TypeMap.PropertyMaps.Where(m =>
				         IsResolveConfigured(m) &&
				         validationContext.TypeMap.ConstructorParameterMatches(m.DestinationName)))
			{
				throw new AutoMapperConfigurationException(
					$"Duplicate mapping in both constructor and member map for {map.TypeMap.DestinationType.FullName} : {map.DestinationName}");
			}
		}
	}

private static bool IsResolveConfigured(MemberMap pm) => (pm.ValueResolverConfig ??
	                                                          pm.CustomMapFunction ?? pm.CustomMapExpression ??
	                                                          (object)pm.ValueConverterConfig) != null;
```

**Reply by lbargaoanu** [reply]
`IsResolveConfigured` is now public in the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build.

**Reply by ChristophHornung** [reply]
@lbargaoanu Thanks for making it public, that makes the unit test more robust. 👍 
Is there any recommendation on what kind of exception to throw during a custom validation?

**Reply by lbargaoanu** [reply]
See [this](https://github.com/AutoMapper/AutoMapper/blob/a00e68fc364f2f8c4987a9c19a45a6b2d7afe998/src/AutoMapper/Configuration/ConfigurationValidator.cs#L127). Your check should probably be `if (validationContext.TypeMap?.ConstructorMapping is true)`.

**Reply by ChristophHornung** [reply]
@lbargaoanu Thanks for the link to throwing exceptions during validation, I will check it out and modify my code.

The `ConstructorMapping` bool is not enough for my use-case as I only want to fail the test if a property is mapped both through the constructor and through a custom mapping (that is why I iterate through the properties map and check each for a match via `m.DestionationName`) . Having a constructor mapping is ok in general.

**Reply by lbargaoanu** [reply]
Just because a `ConstructorMap` exists, it doesn't mean that it's used, `ConstructorMapping` means that it's used indeed. Of course, you can check whatever you feel is right :)

**Reply by ChristophHornung** [reply]
@lbargaoanu  Ah ok, I misunderstood. I will add an additional check for that property so that the check looks like:
```csharp
if (validationContext.TypeMap?.ConstructorMap != null && validationContext.TypeMap.ConstructorMapping)
		{
			foreach (PropertyMap map in validationContext.TypeMap.PropertyMaps.Where(m =>
				         DefaultServerDtoMappingProfileTest.IsResolveConfigured(m) &&
				         validationContext.TypeMap.ConstructorParameterMatches(m.DestinationName)))
			{
				throw new AutoMapperConfigurationException(
					$"Duplicate mapping in both constructor and member map for {map.TypeMap.DestinationType.FullName} : {map.DestinationName}");
			}
		}
```

Just for my understanding, in what case does a constructor mapping exist and not be used?

**Reply by lbargaoanu** [reply]
When it's checked and it doesn't match. But my check _replaces_ yours.

**Reply by ChristophHornung** [reply]
@lbargaoanu I am not sure I understand. What part would it replace?

**Reply by lbargaoanu** [reply]
All you need is `if (validationContext.TypeMap?.ConstructorMapping is true) {...}`.
You can also check how the validator uses `AggreggateException` to throw one exception for multiple errors.

**Reply by ChristophHornung** [reply]
@lbargaoanu 👍  Thanks for the hints, I will look into the validation exceptions.

**Reply by lbargaoanu** [reply]
To simplify usage, I've moved the `ConstructorMapping` check inside `ConstructorParameterMatches`, so it can be removed at the call site.

**Reply by ChristophHornung** [reply]
@lbargaoanu Awesome, thanks! Looking forward to the next nuget release to update my code.

**Reply by ChristophHornung** [reply]
For anyone interested: This is the code updated to 12.0 (the custom `IsResolveConfigured` method is no longer required)

```csharp
private static void ValidateDuplicateMappings(ValidationContext validationContext)
	{
		foreach (PropertyMap map in validationContext.TypeMap?.PropertyMaps.Where(m =>
			                            m.IsResolveConfigured &&
			                            validationContext.TypeMap.ConstructorParameterMatches(m.DestinationName)) ??
		                            new List<PropertyMap>())
		{
			throw new AutoMapperConfigurationException(
				$"Duplicate mapping in both constructor and member map for {map.TypeMap.DestinationType.FullName} : {map.DestinationName}");
		}
	}
```
